### PR TITLE
Add more logs to tail and service status monitor

### PR DESCRIPTION
**Pull request by edwardrf**
And also slight restructure of the service monitor and tail code to improve readability.

### CODE REVIEW

File: src/cmd/cli/command/commands.go [hunk 1]
```diff
@@ -827,26 +826,28 @@ var composeUpCmd = &cobra.Command{
...
+				return // Do not cancel the tailing if we are unable to wait for the service status
 			}
+			cancelTail()
 		}()
-		if err := waitServiceStatus(ctx, cli.ServiceStarted, serviceInfos); err != nil && !errors.Is(err, context.Canceled) {
-			if !errors.Is(err, cli.ErrDryRun) && !errors.As(err, new(cliClient.ErrNotImplemented)) {
-				term.Warnf("failed to wait for service status: %v", err)
+
+		// show users the current streaming logs
+		if err := startTailing(tailCtx, deploy.Etag, since); err != nil {
```
Comment (lionello) [inline src/cmd/cli/command/commands.go:845]: `Tail` will fail when doing `compose up` in the CI (GitHub token has no `tail` scope), but we'd still want to wait for the "up" to complete.

**Review by lionello — CHANGES_REQUESTED** [review]

File: src/cmd/cli/command/commands.go (hunk 1 above)
Comment (edwardrf) [reply src/cmd/cli/command/commands.go:848]: Fixed

File: src/cmd/cli/command/commands.go [hunk 2]
```diff
@@ -825,30 +824,25 @@ var composeUpCmd = &cobra.Command{
...
-			}
-			wg.Wait() // Wait until ctrl+c is pressed
+		// show users the current streaming logs
+		err = startTailing(tailCtx, deploy.Etag, since)
+
+		// Indicate to the user the log tailing failed
+		var cerr *cli.CancelError
+		if !errors.As(err, &cerr) {
+			term.Warnf("failed to start tailing, you will not be able to see logs: %v", err)
 		}
```
Comment (edwardrf) [inline src/cmd/cli/command/commands.go:837]: I am on the fence if this warning is needed

File: src/cmd/cli/command/commands.go [hunk 3]
```diff
@@ -825,30 +824,25 @@ var composeUpCmd = &cobra.Command{
...
-		cancelTail()
-		wg.Wait() // Wait for tail to finish
 
+		// wait for service status to be ready even if tailing failed
+		<-tailCtx.Done()
 		printEndpoints(serviceInfos)
+
+		if err != nil {
+			return err
```
Comment (edwardrf) [inline src/cmd/cli/command/commands.go:844]: This fixes the detach message regression

File: src/cmd/cli/command/commands.go [hunk 4]
```diff
@@ -825,30 +824,25 @@ var composeUpCmd = &cobra.Command{
...
-				term.Warnf("failed to wait for service status: %v", err)
-			}
-			wg.Wait() // Wait until ctrl+c is pressed
+		// show users the current streaming logs
+		err = startTailing(tailCtx, deploy.Etag, since)
+
+		// Indicate to the user the log tailing failed
+		var cerr *cli.CancelError
+		if !errors.As(err, &cerr) {
+			term.Warnf("failed to start tailing, you will not be able to see logs: %v", err)
```
Comment (edwardrf) [inline src/cmd/cli/command/commands.go:836]: I am on the fence if this warning is needed

File: src/cmd/cli/command/servicemonitor.go [hunk 5]
```diff
@@ -2,39 +2,67 @@ package command
...
+		err := waitServiceStatus(ctx, targetStatus, serviceInfos)
+		if err == nil {
+			cancel()
+			return
+		}
+
+		if !errors.Is(err, context.Canceled) && !errors.Is(err, cli.ErrDryRun) && !errors.As(err, new(cliClient.ErrNotImplemented)) {
+			term.Info("Service status monitoring failed, we will continue tailing the logs. Press Ctrl+C to detach.")
+		}
+		term.Debugf("failed to wait for service status: %v", err)
```
Comment (edwardrf) [inline src/cmd/cli/command/servicemonitor.go:27]: The code would be much simpler if we don't need to log these, but we risk swallow errors for debugging

File: src/cmd/cli/command/commands.go [hunk 6]
```diff
@@ -825,30 +824,25 @@ var composeUpCmd = &cobra.Command{
 			return nil
 		}
 
-		tailCtx, cancelTail := context.WithCancel(ctx)
+		tailCtx, cancelTail := contextWithServiceStatus(ctx, cli.ServiceStarted, serviceInfos)
+		defer cancelTail()
```
Comment (lionello) [inline src/cmd/cli/command/commands.go:828]: how can this happen? startTailing blocks? You do `<-Done()` so this cancel is just for cleanup, in which case the name is wrong.

**Pull request closed without merging**